### PR TITLE
Issue42

### PR DESCRIPTION
**Pull request by mjpmorse**
Added an endpoint to keep a list of open games. The table will contain an auto incrementing game id, host name, is open boolean flag, and time of creation. The end point can add, remove, and (get all elements) from the table. Also updated base tables delete method.

### CODE REVIEW

File: quantum_api/src/endpoints/api_endpoint.py [hunk 1]
```diff
@@ -1,15 +1,47 @@
...
 
-    return jsonify(all_games), 200
\ No newline at end of file
+    elif request.method == 'POST':
+        data = request.json
+        host_name = data["host_name"]
+        is_open = bool(data["is_open"])
+        creation_time = data["creation_time"]
+        new_game = PublicGames(host_name=host_name, 
+                               is_open=is_open, creation_time=creation_time)
```
Comment (jtruong04) [inline quantum_api/src/endpoints/api_endpoint.py:33]: Would it be better to have need creation_time? Just have the constructor automatically get the current time when it is called or is that not possible? It just seems like an unnecessary parameter to have the frontend pass in.

File: quantum_api/src/endpoints/api_endpoint.py [hunk 2]
```diff
@@ -1,15 +1,47 @@
...
+        try:
+            PublicGames.post(new_game)
+            return jsonify('Game Added'), 200
```
Comment (jtruong04) [inline quantum_api/src/endpoints/api_endpoint.py:36]: Should probably also return the game itself too.

Comment (mjpmorse) [reply quantum_api/src/endpoints/api_endpoint.py:36]: completed

File: quantum_api/src/endpoints/api_endpoint.py (hunk 1 above)
Comment (mjpmorse) [reply quantum_api/src/endpoints/api_endpoint.py:33]: completed

File: quantum_api/src/endpoints/api_endpoint.py [hunk 3]
```diff
@@ -1,15 +1,47 @@
...
 
 api = Blueprint('api', __name__)
 
-@api.route('/public_games', methods=["GET", "POST"])
+@api.route('/', methods=["GET"])
+def foo():
+    return jsonify('public_games')
+
+@api.route('/public_games', methods=["GET", "POST", "DELETE"])
 def public_game():
```
[the remark this answers is not in the document]
Comment (mjpmorse) [reply quantum_api/src/endpoints/api_endpoint.py:20]: doc string written. returns updated to reflect better practices.

File: quantum_api/src/endpoints/api_endpoint.py [hunk 4]
```diff
@@ -15,33 +18,74 @@ def foo():
...
+        try:
+            all_games = PublicGames.getAll()
+            all_g = []
+            for u in all_games:
+                dictret = dict(u.__dict__)
+                dictret.pop('_sa_instance_state', None)
+                all_g.append(dictret)
+            return jsonify(all_g), 200
+        except Exception as e:
+            return jsonify({"Exemption": str(e)}), 500
```
Comment (jtruong04) [inline quantum_api/src/endpoints/api_endpoint.py:65]: You spelled exception wrong.

**Comment by mjpmorse** [comment]
> Looks good. We'll eventually need a PUT method to toggle is_open once a second player joins. That's primarily the reason I needed the _id returned. But for now, this'll do.

I updated to include this method. Why put off to tomorrow what you can do today.

File: quantum_api/src/endpoints/api_endpoint.py (hunk 4 above)
Comment (mjpmorse) [reply quantum_api/src/endpoints/api_endpoint.py]: fixed it.

File: quantum_api/src/endpoints/api_endpoint.py [hunk 5]
```diff
@@ -16,39 +16,52 @@
 def foo():
     return jsonify('public_games')
 
-@api.route('/public_games', methods=["GET", "POST", "DELETE"])
+@api.route('/public_games', methods=["GET", "POST", "DELETE", "PATCH"])
```
[the remark this answers is not in the document]
Comment (jtruong04) [reply quantum_api/src/endpoints/api_endpoint.py:19]: I also think the endpoint should just be /games. There's never gonna be a /private_games

[the remark this answers is not in the document]
Comment (mjpmorse) [reply quantum_api/src/endpoints/api_endpoint.py]: completed. why can't there exist private games?

[the remark this answers is not in the document]
Comment (jtruong04) [reply quantum_api/src/endpoints/api_endpoint.py]: If we're not saving actual game state data, a game will only ever exist while loaded in the front end. The database exists solely to facilitate people joining public games. Private games will just have the user manually send the game id over to their friend via their favorite communication method. There's never any need to save the game to the database.

If we want, we can save game data to the database. Then people can exit their browser and resume at another time or place or device. Even then, we shouldn't need separate endpoints, just an isPublic property for each game. The endpoint will then just return all public games + private games associated with the user.

**Pull request merged**